colossalai/zero/low_level/low_level_optim.py

```diff
@@ -279,27 +298,73 @@ def _run_reduction(self):
+                        flat_grads = flat_grads.to(self._communication_dtype)
 
                 if not self._partition_grads:
-                    dist.all_reduce(flat_grads, group=self.dp_pg)
-                    if flat_grads.dtype != grad_dtype:
-                        flat_grads = flat_grads.to(grad_dtype)
-
-                    flat_grads_per_rank = flat_grads.split(flat_grads.numel() // self._world_size)
-                    grad_in_bucket = self._bucket_store.get_grad()
-
-                    for rank, grad_list in grad_in_bucket.items():
-                        sync_tensor(flat_grads_per_rank[rank], grad_list)
-                        for grad in grad_list:
-                            param_id = self._bucket_store.get_param_id_of_grad(grad)
-                            if len(self._grad_store.get_partitioned_gradients_by_param_id(group_id,
-                                                                                          param_id)) < self._world_size:
-                                self._grad_store.append_gradients_by_param_id(grad, group_id, param_id)
-                            else:
-                                self._grad_store.add_gradients_by_param_id(grad, rank, group_id, param_id)
+                    if self.extra_dp_pg is None:
+                        dist.all_reduce(flat_grads, group=self.dp_pg)
+                        if flat_grads.dtype != grad_dtype:
+                            flat_grads = flat_grads.to(grad_dtype)
+
+                        flat_grads_per_rank = flat_grads.split(flat_grads.numel() // self._world_size)
+                        grad_in_bucket = self._bucket_store.get_grad()
+
+                        for rank, grad_list in grad_in_bucket.items():
+                            sync_tensor(flat_grads_per_rank[rank], grad_list)
+                            for grad in grad_list:
+                                param_id = self._bucket_store.get_param_id_of_grad(grad)
+                                if len(self._grad_store.get_partitioned_gradients_by_param_id(
+                                        group_id, param_id)) < self._world_size:
+                                    self._grad_store.append_gradients_by_param_id(grad, group_id, param_id)
+                                else:
+                                    self._grad_store.add_gradients_by_param_id(grad, rank, group_id, param_id)
+
+                    # sync extra zero group
+                    else:
+                        # record moe and non moe param
+                        moe_list = []
+                        for param in self._bucket_store._param_list:
+                            moe_list.append(is_moe_tensor(param))
+
+                        # divide them into different groups
+                        moe_grad_list = []
+                        non_moe_grad_list = []
+                        for rank, grad_list in self._bucket_store._grad_in_bucket.items():
```

Comment:
`rank` is not used, why not use `xxx.values()`? By the way, you can also `zip(self._bucket_store._param_list, self._bucket_store._grad_in_bucket.values()`

Reply:
ok, fixed.